Library/Formula/node.rb

```diff
@@ -55,42 +49,54 @@ def install
     system "./configure", *args
     system "make", "install"
 
-    resource("npm").stage libexec/"npm" if build.with? "npm"
-  end
+    if build.with? "npm"
+      resource("npm").stage libexec/"npm_i"
 
-  def post_install
-    return if build.without? "npm"
+      # make sure npm can find node
+      ENV["PATH"] = "#{Formula["node"].prefix}/bin:#{ENV["PATH"]}"
+
+      # set log level temporarily for npm's `make install`
+      ENV["NPM_CONFIG_LOGLEVEL"] = "verbose"
+
+      (libexec/"npm_i").cd { system "./configure", "--prefix=#{libexec}/npm" }
+      (libexec/"npm_i").cd { system "make", "install" }
 
-    (libexec/"npm").cd { system "make", "uninstall" }
-    Pathname.glob(HOMEBREW_PREFIX/"share/man/*") do |man|
-      next unless man.directory?
-      man.children.each do |file|
-        next unless file.symlink?
-        file.unlink if file.readlink.to_s.include? "/node_modules/npm/man/"
+      # Manpages aren't currently linking, which is frustrating.
+      # FIXTHIS before merge, ideally.
+      rm_rf libexec/"npm/lib/node_modules/npm/share"
+
+      if build.with? "completion"
+        bash_completion.install \
+          libexec/"npm_i/lib/utils/completion.sh" => "npm"
       end
     end
+  end
+
+  def post_install
+    return if build.without? "npm"
 
     node_modules = HOMEBREW_PREFIX/"lib/node_modules"
     node_modules.mkpath
-    cp_r libexec/"npm", node_modules
+    npm_exec = node_modules/"npm/bin/npm-cli.js"
+    # Kill npm but preserve all other modules across node updates/upgrades.
+    rm_rf node_modules/"npm" if (node_modules/"npm").exist?
+
+    cp_r libexec/"npm/lib/node_modules/npm", node_modules
+    # This symlink doesn't hop into homebrew_prefix/bin
+    # automatically so remove it and make our own.
+    # This is a small consequence of our bottle npm make install workaround.
+    # All other installs **do** symlink to homebrew_prefix/bin correctly.
+    # We ln rather than cp this because doing so mimics npm's normal install.
+    ln_sf npm_exec, "#{HOMEBREW_PREFIX}/bin/npm"
 
     npm_root = node_modules/"npm"
+
     npmrc = npm_root/"npmrc"
     npmrc.atomic_write("prefix = #{HOMEBREW_PREFIX}\n")
 
-    # set log level temporarily for npm's `make install`
-    ENV["NPM_CONFIG_LOGLEVEL"] = "verbose"
-
-    # make sure npm can find node
-    ENV["PATH"] = "#{opt_bin}:#{ENV["PATH"]}"
+    rm_rf libexec
```

Comment:
Yeh, that was my fault for not being clear. I guess I meant they should be removable but not actually removed. Sorry!